### DISCUSSION TITLE
Automating Setting Exercise Day and Time

### DISCUSSION THREAD

**Discussion by RubyBailey** [jgyates/genmon#1217] [Questions]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Version

V1.19.04

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

Hello,
We have a challenge in that we need to change our generator's exercise weekday regularly, The schedule is available weeks in advance. We could walk to the generator and change it manually but I was thinking about making this more automated. 

I have read through the https://github.com/jgyates/genmon/wiki pages as well as any discussions pertaining to generator exercising. It appears there are the following methods available within genmon to change the generator's exercise schedule:
1. Genmon webpage, i.e. http://ipaddress:8000
2. Genmon.py; monitoring an IMAP email folder, using an email with the subject generator:setexercise=day, time
3. ClientInterface.py; after the ">" prompt, generator:setexercise=day, time

Options 1. and 3. would necessitate a user initiated change, whereas option 2. would allow something like scheduled emails to initiate a more automated change sequence. The challenge with option 2. is that it requires reliable communication to an IMAP email folder. Most of the time that is the case, though it's not always reliable.  Cell service also has reliability challenges.

After looking through the genmon program as well as the above-mentioned options I was thinking about a more "local" software system. Something that could run on the Pi that's running genmon and connected to the generator which would be able to modify the generator exercise schedule based on a locally accessible schedule that is stored on the Pi. 

I was not able to find anything in the existing genmon software library.

The solution I was thinking about was to copy ClientInterface.py, rename the copy, and add a 3rd command line argument (for example -e) that would accept "generator:setexercise=day, time". That new program would then be able send a new exercise weekday and time to genmon.py. (Keeping in mind the limitations of our existing Nexus controller). That python file could be run by a local script, cron timer or other scheduled jobs to modify the generator's exercise schedule. 

Independent of genmon the generator exercise schedule would be updated and stored on the Pi and would not necessitate access to an IMAP email folder.

Some questions: 
1. Your thoughts and comments about the above-mentioned approach.
2. Are there any other options available for sending these commands to genmon.py?

Thanks.

### Additional information

From Genmon Maintenance page:
Maintenance :
Model : Generic Air Cooled
Generator Serial Number : 0007168870
Controller Detected : Nexus, Air Cooled
Nominal RPM : 3600
Rated kW : 20 kW
Nominal Frequency : 60 Hz
Fuel Type : Propane
Generator Phase : 1
Engine Displacement : 999 cc
Controller Settings :
Calibrate Volts : 1009
Nominal Line Voltage : 240 V
Rated Max Power : 20 kW
Exercise :
Exercise Time : Weekly Wednesday 13:02 Quiet Mode On

**Comment by skipfire** [comment]
Genmon is built with a web front end that communicates with APIs that you could call directly.  That specific one calls `http://genmon:8000/cmd/setexercise?setexercise=Monday%2C10%3A00%2CWeekly` to set it to 10:30 on Monday.
You do have to auth the http client before that, but it's a basic form auth so you just make the auth call then the setexercise call right after.  I've written two different things that pull from the genmon APIs for integrating since my needs don't fit other solutions.

**Reply by jgyates** [reply]
making a copy of ClientInterface.py and renaming will also work. As you can see ClientInterface.py is a fairly small python program. It uses the project supplied library ./genmon/genmonlib/myclient.py to send commands to genmon. If it is running in the ./genmon folder then you should have no issues as there are a few dependancies into ./genmon/genmonlib

Let
